# Equipo 1
## Grupo: 001 (Lunes 7:00 - 10:00 pm)
### Integrantes:
+ 1848655 - Rangel Vallejo Isabel Alejandra 
+ 1851167 - Aguilar Sánchez José Antonio
+ 1851124 - Cerecero Peña Angela Scarlett
+ 1851528 - Romero Banda José Manuel
+ 1853849 - García Gutiérrez Alfonso de Jesús 

# Práctica 1: Limpieza de datos

### Título de la base de datos: Wine Reviews
### Url de la página: [Wine Reviews](https://www.kaggle.com/zynicide/wine-reviews?select=winemag-data_first150k.csv)

Para empezar con la limpieza de la base de datos, primero se debe contar con dicha base. Por lo tanto, el primer paso es importar las librerias que nos ayudarán a abrir el archivo

+ Importar librerias

In [1]:
import pandas as pd
import numpy as np

Una vez que se importan las librerias, se puede importar la base de datos llamándola directamente con la función read_csv, debido a que la base está guardada como csv. Es importante mencionar que la base se encuentra guardada en la dirección que tiene este documento, por lo que resulta más fácil leer dicho archivo

+ Importar base de datos

In [2]:
base = pd.read_csv("winemag-data.csv")

La base de datos ahora está guardada en nuestra variable "base", que es un dataframe con los elementos de la base, para visualizarlos se hace uso de distintas funciones

+ Visualizar los datos

Primero se tiene que visualizar toda la base sin modificaciones, con el fin de poder establecer objetivos y saber qué aspectos son los que se tienen que limpiar

In [3]:
base

Unnamed: 0   country  \
0                0     Italy   
1                1  Portugal   
2                2        US   
3                3        US   
4                4        US   
...            ...       ...   
129966      129966   Germany   
129967      129967        US   
129968      129968    France   
129969      129969    France   
129970      129970    France   

                                              description  \
0       Aromas include tropical fruit, broom, brimston...   
1       This is ripe and fruity, a wine that is smooth...   
2       Tart and snappy, the flavors of lime flesh and...   
3       Pineapple rind, lemon pith and orange blossom ...   
4       Much like the regular bottling from 2012, this...   
...                                                   ...   
129966  Notes of honeysuckle and cantaloupe sweeten th...   
129967  Citation is given as much as a decade of bottl...   
129968  Well-drained gravel soil gives this wine its c...   
129969  A dry style of Pinot Gris, this is crisp with ...   
129970  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price  \
0                                 Vulkà Bianco      87    NaN   
1                                     Avidagos      87   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2  \
0       Sicily & Sardinia                 Etna                NaN   
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle  \
0            Kerin O’Keefe          @kerinokeefe   
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 P

Al llamar a la base de datos, podemos ver los elementos que incluye, esto último de forma limitada, debido a que no se carga toda la base y solo muestra algunos elementos de la misma. En algunas ocasiones, tampoco se muestran todas las columnas en caso de que sean muchas y no se alcance a mostrar todas, en nuestro caso sí se muestran todas

Para una mejor perspectiva de lo que tenemos, podemos usar la siguiente función para conocer el número de filas y columnas que tenemos en nuestra base

In [4]:
base.shape

(129971, 14)

Tenemos 129971 filas y 14 columnas en nuestro dataframe

Para una mejor visualización de los datos, podemos hacer uso de la función "head" que nos muestra los primeros "n" elementos de la base, en este caso, pediremos que nos muestre los primeros 5 elementos de la base

In [5]:
base.head(5)

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

Así, podemos conocer un poco de los elementos que conforman nuestra base

De la misma forma, podemos visualizar los últimos "n" elementos del dataframe, con una sencilla función:

In [6]:
base.tail(5)

Unnamed: 0  country  \
129966      129966  Germany   
129967      129967       US   
129968      129968   France   
129969      129969   France   
129970      129970   France   

                                              description  \
129966  Notes of honeysuckle and cantaloupe sweeten th...   
129967  Citation is given as much as a decade of bottl...   
129968  Well-drained gravel soil gives this wine its c...   
129969  A dry style of Pinot Gris, this is crisp with ...   
129970  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price province  \
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0    Mosel   
129967                                     NaN      90   75.0   Oregon   
129968                                   Kritt      90   30.0   Alsace   
129969                                     NaN      90   32.0   Alsace   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   Alsace   

       region_1      region_2         taster_name taster_twitter_handle  \
129966      NaN           NaN  Anna Lee C. Iijima                   NaN   
129967   Oregon  Oregon Other        Paul Gregutt           @paulgwine    
129968   Alsace           NaN          Roger Voss            @vossroger   
129969   Alsace           NaN          Roger Voss            @vossroger   
129970   Alsace           NaN          Roger Voss            @vossroger   

                                                    title         variety  \
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer   

                                          winery  
129966  Dr. H. Thanisch (Erben Müller-Burggraef)  
129967                                  Citation  
129968                           Domaine Gresser  
129969                      Domaine Marcel Deiss  
129970                          Domaine Schoffit

Podemos ver que son los últimos elementos dado el número que tienen al inicio en la columna 1

* Clasificación de los datos

Al observar la base, queda en claro que los datos estan ordenados conforme los obtuvieron, esto nos puede provocar problemas en el futuro para nuestras intenciones sobre este dataframe, por lo que proseguiremos a reacomordarlos de cierta forma para que nos facilite nuestra tarea. En este caso los acomodaremos por pais.

In [7]:
base= base.sort_values('country')

In [8]:
base  # Base ordenada

Unnamed: 0    country  \
90901        90901  Argentina   
122372      122372  Argentina   
122374      122374  Argentina   
4814          4814  Argentina   
4806          4806  Argentina   
...            ...        ...   
124176      124176        NaN   
129407      129407        NaN   
129408      129408        NaN   
129590      129590        NaN   
129900      129900        NaN   

                                              description  \
90901   Odd aromas of bath soap and tropical fruits ar...   
122372  Malty, earthy aromas of cassis and prune are t...   
122374  This SB delivers citrus peel, grass and other ...   
4814    Opens with aromas of tomato leaf, fresh salsa ...   
4806    A heavy, dark and loud Malbec with bullish flo...   
...                                                   ...   
124176  This Swiss red blend is composed of four varie...   
129407  Dry spicy aromas of dusty plum and tomato add ...   
129408  El Capricho is one of Uruguay's more consisten...   
129590  A blend of 60% Syrah, 30% Cabernet Sauvignon a...   
129900  This wine offers a delightful bouquet of black...   

                       designation  points  price          province  \
90901   Valle Las Acequias Clase A      81    9.0  Mendoza Province   
122372                         NaN      86   15.0  Mendoza Province   
122374                 Ciclos Fume      86   28.0             Other   
4814                       Reserva      85   15.0             Other   
4806                           NaN      85   20.0  Mendoza Province   
...                            ...     ...    ...               ...   
124176                Les Romaines      90   30.0               NaN   
129407                     Reserve      89   22.0               NaN   
129408                     Reserve      89   22.0               NaN   
129590                        Shah      90   30.0               NaN   
129900                         NaN      91   32.0               NaN   

          region_1 region_2        taster_name taster_twitter_handle  \
90901      Mendoza      NaN  Michael Schachner           @wineschach   
122372     Mendoza      NaN  Michael Schachner           @wineschach   
122374    Cafayate      NaN  Michael Schachner           @wineschach   
4814       Neuquén      NaN  Michael Schachner           @wineschach   
4806    Uco Valley      NaN  Michael Schachner           @wineschach   
...            ...      ...                ...                   ...   
124176         NaN      NaN       Jeff Jenssen        @worldwineguys   
129407         NaN      NaN  Michael Schachner           @wineschach   
129408         NaN      NaN  Michael Schachner           @wineschach   
129590         NaN      NaN      Mike DeSimone        @worldwineguys   
129900         NaN      NaN      Mike DeSimone        @worldwineguys   

                                                    title             variety  \
90901   Luis Segundo Correas 2015 Valle Las Acequias C...          Chardonnay   
122372  Bodega Familia Barberis 2010 Cabernet Sauvigno...  Cabernet Sauvignon   
122374  Michel Torino 2013 Ciclos Fume Sauvignon Blanc...     Sauvignon Blanc   
4814            Chaltén 2007 Reserva Pinot Noir (Neuquén)          Pinot Noir   
4806                   DiamAndes 2010 Malbec (Uco Valley)              Malbec   
...                                                   ...                 ...   
124176            Les Frères Dutruy 2014 Les Romaines Red           Red Blend   
129407        El Capricho 2015 Reserve Cabernet Sauvignon  Cabernet Sauvignon   
129408               El Capricho 2015 Reserve Tempranillo         Tempranillo   
129590                            Büyülübağ 2012 Shah Red           Red Blend   
129900                                 Psagot 2014 Merlot              Merlot   

                         winery  
90901      Luis Segundo Correas  
122372  Bodega Familia Barberis  
122374            Michel Torino  
4814                    Chaltén  
4806             

Complementando, tambien se puede visualizar las columnas que conforman al dataframe

In [9]:
base.columns

Index(['Unnamed: 0', 'country', 'description', 'designation', 'points',
       'price', 'province', 'region_1', 'region_2', 'taster_name',
       'taster_twitter_handle', 'title', 'variety', 'winery'],
      dtype='object')

A simple vista podemos apreciar que la primera columna no tiene nombre, por lo que sabemos que más adelante nos ocasionará problemas. Del mismo modo, algunos nombres de columna son muy largos o no definen correctamente los elementos de la columna. Para evitar estos problemas, se puede redefinir el nombre de las columnas

+ Cambiar nombres de columnas

In [10]:
nueva_base = base.rename(columns = {'Unnamed: 0':'Item',
                       'taster_twitter_handle':'taster_twitter'})
# se cambia el nombre de la primera columna debido a que no contaba con un numbre que hiciera mención al numero de elemento que es
# se cambia el nombre de la columna 'taster_twitter_handle' ya que se considera muy largo. Se opta por acortarlo

In [11]:
nueva_base

Item    country                                        description  \
90901    90901  Argentina  Odd aromas of bath soap and tropical fruits ar...   
122372  122372  Argentina  Malty, earthy aromas of cassis and prune are t...   
122374  122374  Argentina  This SB delivers citrus peel, grass and other ...   
4814      4814  Argentina  Opens with aromas of tomato leaf, fresh salsa ...   
4806      4806  Argentina  A heavy, dark and loud Malbec with bullish flo...   
...        ...        ...                                                ...   
124176  124176        NaN  This Swiss red blend is composed of four varie...   
129407  129407        NaN  Dry spicy aromas of dusty plum and tomato add ...   
129408  129408        NaN  El Capricho is one of Uruguay's more consisten...   
129590  129590        NaN  A blend of 60% Syrah, 30% Cabernet Sauvignon a...   
129900  129900        NaN  This wine offers a delightful bouquet of black...   

                       designation  points  price          province  \
90901   Valle Las Acequias Clase A      81    9.0  Mendoza Province   
122372                         NaN      86   15.0  Mendoza Province   
122374                 Ciclos Fume      86   28.0             Other   
4814                       Reserva      85   15.0             Other   
4806                           NaN      85   20.0  Mendoza Province   
...                            ...     ...    ...               ...   
124176                Les Romaines      90   30.0               NaN   
129407                     Reserve      89   22.0               NaN   
129408                     Reserve      89   22.0               NaN   
129590                        Shah      90   30.0               NaN   
129900                         NaN      91   32.0               NaN   

          region_1 region_2        taster_name  taster_twitter  \
90901      Mendoza      NaN  Michael Schachner     @wineschach   
122372     Mendoza      NaN  Michael Schachner     @wineschach   
122374    Cafayate      NaN  Michael Schachner     @wineschach   
4814       Neuquén      NaN  Michael Schachner     @wineschach   
4806    Uco Valley      NaN  Michael Schachner     @wineschach   
...            ...      ...                ...             ...   
124176         NaN      NaN       Jeff Jenssen  @worldwineguys   
129407         NaN      NaN  Michael Schachner     @wineschach   
129408         NaN      NaN  Michael Schachner     @wineschach   
129590         NaN      NaN      Mike DeSimone  @worldwineguys   
129900         NaN      NaN      Mike DeSimone  @worldwineguys   

                                                    title             variety  \
90901   Luis Segundo Correas 2015 Valle Las Acequias C...          Chardonnay   
122372  Bodega Familia Barberis 2010 Cabernet Sauvigno...  Cabernet Sauvignon   
122374  Michel Torino 2013 Ciclos Fume Sauvignon Blanc...     Sauvignon Blanc   
4814            Chaltén 2007 Reserva Pinot Noir (Neuquén)          Pinot Noir   
4806                   DiamAndes 2010 Malbec (Uco Valley)              Malbec   
...                                                   ...                 ...   
124176            Les Frères Dutruy 2014 Les Romaines Red           Red Blend   
129407        El Capricho 2015 Reserve Cabernet Sauvignon  Cabernet Sauvignon   
129408               El Capricho 2015 Reserve Tempranillo         Tempranillo   
129590                            Büyülübağ 2012 Shah Red           Red Blend   
129900                                 Psagot 2014 Merlot              Merlot   

                         winery  
90901      Luis Segundo Correas  
122372  Bodega Familia Barberis  
122374            Michel Torino  
4814                    Chaltén  
4806                  DiamAndes  
...                         ...  
124176        Les Frères Dutruy  
129407              El Capricho  
129408              El Capricho  
129590                Büyülübağ  
129900                   Psagot  

[129971 rows x 14 col

Con la última impresión, podemos notar que los nombres de las columnas se han cambiado satisfactoriamente

Una vez que los nombres de las columnas son correctos, se puede estudiar a los elementos de las columnas

+ Descripción básica de columnas

Tipos de datos en las columnas:

In [12]:
nueva_base.dtypes

Item                int64
country            object
description        object
designation        object
points              int64
price             float64
province           object
region_1           object
region_2           object
taster_name        object
taster_twitter     object
title              object
variety            object
winery             object
dtype: object

Con esta función podemos analizar los tipos de datos que hay en cada columna. Para este caso, la mayoría son de tipo object, lo cual tiene sentido debido a que los únicos datos numericos son: número de item, puntos recibidos y precio de la botella. Por lo que los tipos de datos de las columnas son correctos

Si se requiere analizar a los elementos en cuestión de una columna, podemos hacer uso de la función "unique", la cual nos regresa los valores únicos que existen en la columna, es decir, sin repetición

In [13]:
nueva_base.country.unique()

array(['Argentina', 'Armenia', 'Australia', 'Austria',
       'Bosnia and Herzegovina', 'Brazil', 'Bulgaria', 'Canada', 'Chile',
       'China', 'Croatia', 'Cyprus', 'Czech Republic', 'Egypt', 'England',
       'France', 'Georgia', 'Germany', 'Greece', 'Hungary', 'India',
       'Israel', 'Italy', 'Lebanon', 'Luxembourg', 'Macedonia', 'Mexico',
       'Moldova', 'Morocco', 'New Zealand', 'Peru', 'Portugal', 'Romania',
       'Serbia', 'Slovakia', 'Slovenia', 'South Africa', 'Spain',
       'Switzerland', 'Turkey', 'US', 'Ukraine', 'Uruguay', nan],
      dtype=object)

En este último caso apreciamos a los países que aparecen en nuestra base de datos

In [14]:
nueva_base.points.unique()

array([ 81,  86,  85,  84,  83,  90,  82,  87,  89,  88,  91,  80,  93,
        92,  95,  94,  97,  96,  99,  98, 100], dtype=int64)

Para este caso, observamos todas las calificaciones que fueron dadas a los vinos. Sin embargo, no por conocer todas las calificaciones que se dieron sabemos a quiénes se les dio esta calificación

También se tiene que tener cuidado al utilizar esta operación, debido a que no todas las columnas tienen elementos repetidos. En nuestra base, la columna "Item" no tiene elementos repetidos, por lo que tendríamos una serie de números enlistados del 1 hasta el elemento final de la base. Por lo que se debería usar cuando queremos agrupar datos o categorizarlos; o simplemente conocer los valores con los que contamos

Aunado a esto, es de vital importancia conocer si nuestra base cuenta con datos nulos, por lo que podemos utilizar la siguiente función:

In [15]:
nueva_base.isnull().any()

Item              False
country            True
description       False
designation        True
points            False
price              True
province           True
region_1           True
region_2           True
taster_name        True
taster_twitter     True
title             False
variety            True
winery            False
dtype: bool

La cual nos regresa "False" si la columna no tiene datos nulos, caso contrario, nos regresa "True" si la columna cuenta con datos nulos. Podemos observar que la mayoría de las columnas tiene por lo menos un dato nulo

Otra forma de conocer con certeza los datos nulos en la base, es con la siguiente función:

In [16]:
nueva_base.isnull().sum()

Item                  0
country              63
description           0
designation       37465
points                0
price              8996
province             63
region_1          21247
region_2          79460
taster_name       26244
taster_twitter    31213
title                 0
variety               1
winery                0
dtype: int64

Retorna el número de datos nulos que contiene cada columna. Esto ayuda mucho al momento de tomar decisiones, ya que se podría eliminar una columna que contenga más datos nulos que elementos en sí. Aquí se deben de tomar decisiones de qué es lo que se hará con los datos nulos

Una de las decisiones que se debe de tomar es el de eliminar las columnas que no se vayan a utilizar, debido a que ocupan espacio al momento de la visualización y la memoria del equipo. Por lo que nos debemos plantear la incógnita de qué columnas necesitamos verdaderamente en base a nuestros objetivos. Para recordar las columnas que tenemos, haremos lo siguiente:

In [17]:
nueva_base.columns

Index(['Item', 'country', 'description', 'designation', 'points', 'price',
       'province', 'region_1', 'region_2', 'taster_name', 'taster_twitter',
       'title', 'variety', 'winery'],
      dtype='object')

Si no se sabe cuáles columnas eliminar, se puede elaborar un análisis para conocer las más importantes o vitales para la investigación. En nuestro caso, la columna "Item" es importante porque nos permite la correcta indexación de la base; la columna "country" es importante porque nos cuenta el origen del vino; la columna "points" nos dice la calificación del vino; la columna "price" nos dice el precio del vino. Podríamos afirmar que estas columnas son de vital importancia.

En un segundo plano podrían estar las columnas "taster_name" debido a que nos dicta el nombre del crítico; "title" ya que nos da el nombre del vino; "description" porque dice una descripción de cómo es el vino y sus componentes. Así que podemos eliminar las columnas que no se mencionan, debido a que no se encuentra un serio uso a estas.

+ Eliminación de columnas no deseadas

In [18]:
edit_base = nueva_base.drop(['designation', 'province', 'region_1', 'region_2', 'taster_twitter', 'winery'], axis=1)

In [19]:
edit_base

Item    country                                        description  \
90901    90901  Argentina  Odd aromas of bath soap and tropical fruits ar...   
122372  122372  Argentina  Malty, earthy aromas of cassis and prune are t...   
122374  122374  Argentina  This SB delivers citrus peel, grass and other ...   
4814      4814  Argentina  Opens with aromas of tomato leaf, fresh salsa ...   
4806      4806  Argentina  A heavy, dark and loud Malbec with bullish flo...   
...        ...        ...                                                ...   
124176  124176        NaN  This Swiss red blend is composed of four varie...   
129407  129407        NaN  Dry spicy aromas of dusty plum and tomato add ...   
129408  129408        NaN  El Capricho is one of Uruguay's more consisten...   
129590  129590        NaN  A blend of 60% Syrah, 30% Cabernet Sauvignon a...   
129900  129900        NaN  This wine offers a delightful bouquet of black...   

        points  price        taster_name  \
90901       81    9.0  Michael Schachner   
122372      86   15.0  Michael Schachner   
122374      86   28.0  Michael Schachner   
4814        85   15.0  Michael Schachner   
4806        85   20.0  Michael Schachner   
...        ...    ...                ...   
124176      90   30.0       Jeff Jenssen   
129407      89   22.0  Michael Schachner   
129408      89   22.0  Michael Schachner   
129590      90   30.0      Mike DeSimone   
129900      91   32.0      Mike DeSimone   

                                                    title             variety  
90901   Luis Segundo Correas 2015 Valle Las Acequias C...          Chardonnay  
122372  Bodega Familia Barberis 2010 Cabernet Sauvigno...  Cabernet Sauvignon  
122374  Michel Torino 2013 Ciclos Fume Sauvignon Blanc...     Sauvignon Blanc  
4814            Chaltén 2007 Reserva Pinot Noir (Neuquén)          Pinot Noir  
4806                   DiamAndes 2010 Malbec (Uco Valley)              Malbec  
...                                                   ...                 ...  
124176            Les Frères Dutruy 2014 Les Romaines Red           Red Blend  
129407        El Capricho 2015 Reserve Cabernet Sauvignon  Cabernet Sauvignon  
129408               El Capricho 2015 Reserve Tempranillo         Tempranillo  
129590                            Büyülübağ 2012 Shah Red           Red Blend  
129900                                 Psagot 2014 Merlot              Merlot  

[129971 rows x 8 columns]

Podemos ver que las columnas se han eliminado satisfactoriamente

Una vez que tenemos solamente las columnas que tenemos, nos preguntamos: ¿Qué haremos con los datos nulos que quedaron?. Para responder esto, hay que estudiar los datos que tenemos. De nueva cuenta veremos los datos nulos que tenemos.

In [20]:
edit_base.isnull().sum()

Item               0
country           63
description        0
points             0
price           8996
taster_name    26244
title              0
variety            1
dtype: int64

Sin duda alguna, al enfocar nuestro trabajo en la variable "price", es un elemento vital. Por lo que se optará por eliminar todas las filas que en la columna "price" tengan un dato nulo, debido a que pueden interferir con el estudio que se desea hacer. Por obvias razones, no se les puede rellenar con 0, ya que estaríamos diciendo que el precio de la botella de vino es de cero, cuando no lo es

Sin duda alguna, ya que nuestro trabajo se enfoca en la variable "price", debemos eliminar aquellas filas cuyo valor en "price" sea nulo. Primero llenaremos los datos nulos con 0, para posteriormente filtrar de acuerdo a este elemento y no contemplar los items que su valor en "price" sea 0, ya que sabremos que proviene de un dato nulo o que, en dado caso, no puede ser posible

In [21]:
df = edit_base.fillna(0)

In [22]:
df.isnull().sum()

Item           0
country        0
description    0
points         0
price          0
taster_name    0
title          0
variety        0
dtype: int64

Ya no tenemos datos nulos, han sido rellenados con cero. Procedemos a filtrar por este valor

In [23]:
base_resultante = df[df['price'] != 0]

In [24]:
base_resultante.isnull().sum()

Item           0
country        0
description    0
points         0
price          0
taster_name    0
title          0
variety        0
dtype: int64

Podemos notar que al eliminar los datos nulos en la columna de precio, nos quedamos sin datos nulos en toda la demás base. Pero, ¿con cuántos elementos contamos ahora?

In [25]:
base_resultante.shape

(120975, 8)

Aún contamos con 120975 datos, los cuales son un buen número para continuar.

Para una mejor compresión de los datos que tenemos, podemos agrupar los datos de acuerdo a categorías. Si recordamos bien, la base tiene una columna llamada "country" por lo que es una buena opción agrupar los datos por país.

In [26]:
base_resultante.country.unique()

array(['Argentina', 'Armenia', 'Australia', 'Austria',
       'Bosnia and Herzegovina', 'Brazil', 'Bulgaria', 'Canada', 'Chile',
       'China', 'Croatia', 'Cyprus', 'Czech Republic', 'England',
       'France', 'Georgia', 'Germany', 'Greece', 'Hungary', 'India',
       'Israel', 'Italy', 'Lebanon', 'Luxembourg', 'Macedonia', 'Mexico',
       'Moldova', 'Morocco', 'New Zealand', 'Peru', 'Portugal', 'Romania',
       'Serbia', 'Slovakia', 'Slovenia', 'South Africa', 'Spain',
       'Switzerland', 'Turkey', 'US', 'Ukraine', 'Uruguay', 0],
      dtype=object)

In [27]:
# Si agrupamos por Estados Unidos
EUA = base_resultante[base_resultante['country']=='US']

EUA.to_csv('Solo_EUA.csv', index=False, sep = ',')

In [28]:
# Si agrupamos por España
Espania = base_resultante[base_resultante['country']=='Spain']

Espania.to_csv('Solo_Espania.csv', index=False, sep = ',')

In [29]:
# Si agrupamos por Italia
Italia = base_resultante[base_resultante['country']=='Italy']

Italia.to_csv('Solo_Italia.csv', index=False, sep = ',')

De esta misma manera, podríamos agrupar en base a los puntos obtenidos. Los cuales son:

In [30]:
base_resultante.points.unique()

array([ 81,  86,  85,  84,  83,  90,  82,  87,  89,  88,  91,  80,  93,
        92,  95,  94,  97,  96,  99,  98, 100], dtype=int64)

In [31]:
# Agrupando por puntos obtenidos
# Si agrupamos por España
Vinos_100 = base_resultante[base_resultante['points']== 100]

Vinos_100.to_csv('Solo_Vinos_100.csv', index=False, sep = ',')

## Preguntas de interés

+ ¿Para qué me puede servir estas agrupaciones para la práctica siguiente?

**R:** La categorización de los elementos nos ayuda a identificar características de ellos más fácilmente, así mismo, nos permite establecer relaciones entre los datos para poder predecir comportamientos o patrones entre ellos. También, al agrupar, podemos darnos cuenta de los datos que quedan sin agrupar, y aquí se podría tomar la decisión de incluirlos o no en nuestra investigación.  
El equipo trabajará con una base de datos que cuenta con una columna con los nombres de los países, así que si quisieramos establecer relaciones, primero deberíamos voltear a ver a los países por separado, por lo que preveemos que las agrupaciones de datos nos serán de utilidad en nuestros futuros trabajos.



+ Importancia del análisis previo de los datos antes de iniciar las modificaciones

**R:** Es importante analizar los datos de nuestra base antes de comenzar con las modificaciones del dataframe, debido a que debemos conocer a la perfección cómo es que están conformados; es decir, con qué variables contamos, qué tipo de datos tenemos, qué podemos hacer con los datos, entre otras preguntas que hay que contestar antes de empezar a trabajarlos. Es de vital importancia conocer nuestras limitaciones para poder enfocar nuestro tiempo en las cosas que sí podemos hacer. De la misma manera, saber cuáles son nuestras variables nos ayuda a establecer objetivos preeliminares antes de comenzar, para no empezar sin un rumbo e ir encaminados hacia algo. Es importante tener en mente lo que deseamos hacer con la base o nuestro objeto de estudio, para, de primera mano, poder descartar elementos que no nos ayudarán a alcanzar los objetivos. Conocer la naturaleza de los datos también es importante, ya que debemos considerarlos al momento de realizar las modificaciones, así que primero deberíamos saber cómo están conformados nuestros datos.

## Tareas extras:

+ Investigar los siguientes comandos: ?df.dropna() ?df.astype

**df.dropna()**


Función que se utiliza para eliminar renglones y columnas que contengan elementos nulos.
La sintaxis completa de esta función queda de la siguiente manera:  
dataframe.dropna(axis=0, how = 'any', thresh = None, subset = None, inplace = False)  
Donde:
+ **axis**: Acepta los valores de 0 o 1 (dependiendo el caso). Con 0 se eliminan los renglones que contengan valores nulos; con 1, elimina las columnas que contengan valores nulos. Por default es 0.   
+ **how**: Acepta {'any', 'all'}. Con 'any' eliminará todas las columnas (o renglones) que tengan al menos un elemento nulo. Por otro lado, 'all' eliminará las columnas (o renglones) que tengan todos sus elementos nulos. Por default es 'any'.  
+ **thresh**: Valor entero, opcional. Se menciona al menos cuántos elementos no nulos debe tener una columna (o renglón) para mantenerse.  
+ **subset**: Arreglo, opcional. Se mencionan las columnas (o renglones) en los cuales se desea buscar información. Si tú estás eliminando renglones, este arreglo será de nombres de columnas y viceversa.  
+ **inplace**: booleano, default False. Si es verdadero, hace una operación en su lugar.

Para mejor comprensión de la función se realizará una serie de ejemplos:

In [32]:
# se crea un dataframe:
dataf = pd.DataFrame({"nombre": ['Alfonso', 'Angela', 'Jose', 'Isabel', 'Manuel'],
                      "edad": [20, np.nan, 19, 20, np.nan],
                      "calificacion": [np.nan, 98, 100, 99, 97]})
dataf

nombre  edad  calificacion
0  Alfonso  20.0           NaN
1   Angela   NaN          98.0
2     Jose  19.0         100.0
3   Isabel  20.0          99.0
4   Manuel   NaN          97.0

In [33]:
# eliminar los renglones donde al menos un elemento falta
dataf.dropna()

nombre  edad  calificacion
2    Jose  19.0         100.0
3  Isabel  20.0          99.0

In [34]:
# eliminar las columnas donde al menos un elemento falta
dataf.dropna(axis='columns')

nombre
0  Alfonso
1   Angela
2     Jose
3   Isabel
4   Manuel

In [35]:
# eliminar los renglones donde todos los elementos faltan
dataf.dropna(how = 'all')

nombre  edad  calificacion
0  Alfonso  20.0           NaN
1   Angela   NaN          98.0
2     Jose  19.0         100.0
3   Isabel  20.0          99.0
4   Manuel   NaN          97.0

In [36]:
# Quédese solo con los renglones con al menos 2 elementos no nulos
dataf.dropna(thresh = 2)

nombre  edad  calificacion
0  Alfonso  20.0           NaN
1   Angela   NaN          98.0
2     Jose  19.0         100.0
3   Isabel  20.0          99.0
4   Manuel   NaN          97.0

**Fuente consultada**: [recurso utilizado](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html)

**df.astype()**

Función utilizada para cambiar el tipo de dato de las columnas, se puede cambiar una o varias a la vez (se necesita un diccionario para esto). La sintaxis completa queda de la siguiente manera:
dataframe.astype(dtype, copy = True, errors = 'raise')  
Donde:  
+ **dtype**: Tipo de datos que queremos asignar a nuestro objeto.
+ **copy**: Un parámetro booleano. Devuelve una copia cuando tenemos True.
+ **errors**: Controla el lanzamiento de excepciones sobre datos no válidos para el tipo de datos proporcionado. Tiene dos opciones:
     + raise: permite que se levanten las excepciones
     + inore: suprime las excepciones. Si existe un error, entonces devuelve el objeto original.    
     
Para comprensión del tema, se elaborarán ejemplos sencillos del uso de la función.

In [37]:
import pandas as pd

dataframe=pd.DataFrame({'Asistencia': {0: 60, 1: 100, 2: 80,3: 78,4: 95},
                        'Nombre': {0: 'Olivia', 1: 'John', 2: 'Laura',3: 'Ben',4: 'Kevin'},
                        'Marcas obtenidas': {0: 90, 1: 75, 2: 82, 3: 64, 4: 45}})

# Vemos nuestro dataframe
dataframe

Asistencia  Nombre  Marcas obtenidas
0          60  Olivia                90
1         100    John                75
2          80   Laura                82
3          78     Ben                64
4          95   Kevin                45

In [38]:
print("Los tipos originales de los datos de nuestro data frome son: \n ")
print(dataframe.dtypes)

Los tipos originales de los datos de nuestro data frome son: 
 
Asistencia           int64
Nombre              object
Marcas obtenidas     int64
dtype: object


In [39]:
# Ahora utilizaremos la función para cambiar el tipo de datos de la columna de asistencia

dataframe1 = dataframe.astype({'Asistencia': 'int32'})
print("Los tipos de datos de nuestro dataframe modificado son: \n")
print(dataframe1.dtypes)

Los tipos de datos de nuestro dataframe modificado son: 

Asistencia           int32
Nombre              object
Marcas obtenidas     int64
dtype: object


In [40]:
# Ahora se cambiarán los tipos de datos de todas las columnas del dataframe
dataframe2 = dataframe.astype('object')
print("Los tipos de datos de nuestro dataframe modificado completamente son: \n")
print(dataframe2.dtypes)

Los tipos de datos de nuestro dataframe modificado completamente son: 

Asistencia          object
Nombre              object
Marcas obtenidas    object
dtype: object


In [41]:
# Cambiaremos el tipo de datos de la columna 'Marcas obtenidas' a tipo string
dataframe3 = dataframe.astype({'Marcas obtenidas': 'str'})
print("Los tipos de datos de nuestro dataframe modificado completamente son: \n")
print(dataframe3.dtypes)

Los tipos de datos de nuestro dataframe modificado completamente son: 

Asistencia           int64
Nombre              object
Marcas obtenidas    object
dtype: object


**Fuente consultada:** [recurso utilizado](https://www.delftstack.com/es/api/python-pandas/pandas-dataframe-dataframe.astype-function/)



+ Realizar una tarea de preprocesamiento no mencionada en esta práctica.

## Otra función para agrupar

Otra función que sirve para agrupar los datos que tenemos, es groupby. Groupby agrupa nuestros datos de acuerdo a las categorias que queremos. Del ejemplo anterior, pudimos notar que los paises son los que más se repiten, por lo que agrupar por país serviría para visualizar los datos de una mejor manera. Groupby se utiliza de la siguiente manera:

In [42]:
base_agrupada = base_resultante.groupby('country')

In [43]:
print(base_agrupada['price'].mean())

country
0                         24.593220
Argentina                 24.510117
Armenia                   14.500000
Australia                 35.437663
Austria                   30.762772
Bosnia and Herzegovina    12.500000
Brazil                    23.765957
Bulgaria                  14.645390
Canada                    35.712598
Chile                     20.786458
China                     18.000000
Croatia                   25.450704
Cyprus                    16.272727
Czech Republic            24.250000
England                   51.681159
France                    41.139120
Georgia                   19.321429
Germany                   42.257547
Greece                    22.364425
Hungary                   40.648276
India                     13.333333
Israel                    31.768916
Italy                     39.663770
Lebanon                   30.685714
Luxembourg                23.333333
Macedonia                 15.583333
Mexico                    26.785714
Moldova             

De esta manera, contabilizamos de una manera más rápida caracteristicas básicas de los elementos de nuestra base. Además de su promedio, se puede ver el valor mínimo, máximo, su desviación estandar, entre otras operaciones. Podemos notar que es una operación sencilla y eficiente

In [44]:
print(base_agrupada['price'].max())

country
0                           50.0
Argentina                  230.0
Armenia                     15.0
Australia                  850.0
Austria                   1100.0
Bosnia and Herzegovina      13.0
Brazil                      60.0
Bulgaria                   100.0
Canada                     120.0
Chile                      400.0
China                       18.0
Croatia                     65.0
Cyprus                      21.0
Czech Republic              45.0
England                     95.0
France                    3300.0
Georgia                     40.0
Germany                    775.0
Greece                      79.0
Hungary                    764.0
India                       20.0
Israel                     150.0
Italy                      900.0
Lebanon                     75.0
Luxembourg                  30.0
Macedonia                   20.0
Mexico                     108.0
Moldova                     42.0
Morocco                     40.0
New Zealand                130.0
Pe

In [45]:
print(base_agrupada['price'].min())

country
0                          6.0
Argentina                  4.0
Armenia                   14.0
Australia                  5.0
Austria                    7.0
Bosnia and Herzegovina    12.0
Brazil                    10.0
Bulgaria                   8.0
Canada                    12.0
Chile                      5.0
China                     18.0
Croatia                   12.0
Cyprus                    11.0
Czech Republic            15.0
England                   25.0
France                     5.0
Georgia                    9.0
Germany                    5.0
Greece                     8.0
Hungary                   10.0
India                     10.0
Israel                     8.0
Italy                      5.0
Lebanon                   13.0
Luxembourg                16.0
Macedonia                 15.0
Mexico                     8.0
Moldova                    8.0
Morocco                   14.0
New Zealand                7.0
Peru                      10.0
Portugal                   5.0
